README
Revinetd version 1.0.1

Revinetd is the GNU Version of the TCP gender changer.

Portions Copyright (C) 2003-2008 Steven M. Gill
Portions Copyright (C) 2003-2004 Alexandre Carmel-Veilleux

The getopt.c and getopt.h files included in this distributions are
borrowed from the GNU libc distributions and are included to
provide long arguments functionality to operating systems that
do not include it. They are GPL licensed under the same term as
this software and are copyrighten to The Free Software Foundation.

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, 
USA.

This program can be used to proxy information from one port to 
another using two modes:

- Server Mode (Listen-Listen)
- Relay Agent Mode (Connect-Connect)

The main reason for using this program is for situations when the 
established connection path is the opposite from the data flow.  Typically 
most firewalls are configured in a more insecure manner for outbound 
traffic then inbound.  If we want to connect to a service (e.g  NetBIOS) 
that is filtered by a firewall, we can utilize a Relay Agent Mode 
revinetd instance to connect to the target service and outbound through 
the firewall's more liberal filtering rules.  On the other side, a 
Server Mode revinetd instance will accept a connection from the 
client computer and the connect-connect mode instance.  The picture below 
shows the connection and data flow:

Data Flow: (Client initiated traffic such as NetBIOS or VNC)

     (Outside / Internet)      Firewall          (Inside)
                                  |
Client ---> Server Mode ----------|---------> Relay Agent Mode ---> Service
                                  |


Connection Flow:

     (Outside / Internet)      Firewall          (Inside)
                                  |
Client ---> Server Mode <---------|---------- Relay Agent Mode ---> Service
                                  |

As you can see the Relay Agent Mode instance initiates the connection back
to the Server Mode instance through the firewall.  Since most firewalls are
configured with a more liberal rule set on the inside, there are typically
more options to connect out with then connect in.  The trick is gaining
access to install the Relay Agent Mode instance on the inside.

You may say, "What's the point?" or "This sounds a lot like what a hacker 
will do!" The answer is yes, that's what a hacker could or would do.  

The idea is already out there as is stated in Compass Security's excellent 
paper entitled "TCP/IP Gender Changer" and can be located at:

http://www.csnc.ch/downloads/docs/techdocs/TCP-IP_GenderChanger_CSNC_V1.0.pdf

What we are doing here is providing the tool to penetration testers to 
show why it is just as important to secure a firewall from connectivity on 
both sides.  There are a lot of good pen testers out there who just don't 
have the time to write these programs and we are providing this to them.  
I do NOT, however, condone any use of this product for any malicous 
activity.  We are merely providing an easy-to-use connectivity tool to 
pen testers to help get the point accross.

Though I have no evidence, I am sure that there are most likely many 
hackers who whave used netcat or the like with some shell scripts to 
accomplish the same task. 

USAGE:

Server Mode

revinetd -s -c <address1:port1> -l <address2:port2> 

-l tells Server mode where the relay agent will connect and -c tells the
server where the client will connect.  Of course on  *nix  systems,
anything below 1024 needs root access.

Relay Agent Mode

revinetd -r -t <host1:port1> -b <host2:port2>

-t is the target to connect to (ultimate destination), and -b is the
server callback port (peer which has -l listening).  Specify host:port
format for both (e.g. localhost:1234 or 127.0.0.1:1234).

NOTE:  This has only been tested on the following configurations:

Linux 2.4,2.6
OpenBSD 2.9+
FreeBSD 4.4+

Cygwin with gcc 3.2.2 ** Do not connect to loopback... use the network 
adapter's address instead.

revinetd now compiles under solaris without monkeying around with the
makefile, but still needs testing!

Please feel free to send us your successes or any pathces need to get
revinetd working on your system.



ChangeLog
8/27/2008
Added bugfix for broken relase of relay agent.

12/9/2004
Added ability to chose the ip address to bind to with server mode.
This is ready for 1.0.1 release.


9/x/2004
Fixed connection to not require name resolution.  I encountered an older
server where this was broke.  inet_aton was also not available on this 
server, so we test for it on configure and use an older connection 
method.  

4/27/2004
Updated to 1.0

Hearbeat works and is tested
Makefile includes install ad behave with PREFIX chosen.

9/29/2003
Fixed keep alive.  Keep alive breaks when timeout occurs when select is consitently
thrown only relay agent port pairs.  It stays at a three minute timeout instead
of what is configured.

6/27/2003

Keep-alive is now a configurable option for the relay agent.  That way
if we have have a firewall with a quick timout on the connection 
tracking, we can keep the comm channel from going stale.

4/03/2003

Seem to have working keep-alive.

3/21/2003

Server: when relay agent disconnects, server resets state.

3/19/2003

Updated configure.in to correctly compile under Solaris SPARC under cc
and gcc

3/14/2003

Remove references to mode.  OOB works so nice that this is how we intend
to operate only.

3/9/2003

The whole of the network code has been rewritten from scratch. It's
cleaner and it handles multiple concurrent connections, like web
browsing. Only -m out works, however.

2/13/2003

Major changes. A new switch has been added.  Mode (-m) will specify the
communication mode.  currently this is MODE_NONE which gives us crappy
connectivity issues like before, but still only uses one connection.  -m
out is Out-Of-Band (OOB) communication channel.  This allows us to have
control variables about client and target connections, so we can
simulate a forward data tunnel connect.  This is done through a second
connection which is the comm channel.  This connection is also initiated
from the relay agent and connects on the same port as the data tunnel.
This is so we still only need one port on the firewall open.  The
"protocol" on the data channel is very simple, it only communicates the
state of the connection on each side.  For now, specify -m out to enable
this.  The "none" mode is default, but will either go away soon, or
become an depereciated option for old compatibility (though this isn't
really old).

1/2/2003

Updated License Files and Readme

Updated copy_between_ports function for 2k buffered copy


1/1/2003

Initial Code Release


revinetd.c
/*
 * revinetd.c
 *
 * This file is a part of the revinetd project              
 *
 * Revinetd is copyright (c) 2003-2008 by Steven M. Gill
 * and distributed under the GPL.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307,
 * USA.
 *
 *                                                                        */

#include "includes.h"
#include "revinetd.h"
#include "relay_agt.h"
#include "server.h"
#include "misc.h"

static const char cvsid[] = "$Id: revinetd.c,v 1.29 2008/08/28 03:24:59 necrotaur Exp $";

int
main(int argc, char **argv)
{
    /* All variables are initialized to a known invalid value so that they
       are not used un-initialized. To use them should trip tests or cause
       segfaults. */
    static struct option long_options[] = {
        /* These options set a flag. */
        {"help", no_argument, 0, 'h'},
        {"daemonize", no_argument, 0, 'd'},
        {"relay-agent", no_argument, 0, 'r'},
        {"server", no_argument, 0, 's'},
        {"listen-client", required_argument, 0, 'c'},
        {"listen-relay", required_argument, 0, 'l'},
        {"server-host", required_argument, 0, 'b'},
        {"target-host", required_argument, 0, 't'},
        {"verbose", no_argument, 0, 'v'},
        {"quiet", no_argument, 0, 'q'},
        {"keep-alive", required_argument, 0, 'k'},
        {0, 0, 0, 0}};
    /* getopt_long stores the option index here. */
    int c, option_index = 0;
    pid_t pid;
   
    printf("%s - Copyright (c) 2003-2008 Steven M. Gill\n", *argv);
    printf("%s is distributed under the terms of the GPL\n", *argv);
    printf("http://revinetd.sourceforge.net\n\n");
    exec_name = strdup(*argv);
    
    init_conf();

    while (1) {
        c = getopt_long(argc, argv, "hdrsc:l:b:t:qvk:", long_options,
                &option_index);

        /* Detect the end of the options. */
        if (c == -1)
            break;

        switch (c) {
        case 0:
            /* If this option set a flag, do nothing else now. */
            if (long_options[option_index].flag != 0)
                break;
            printf("option %s", long_options[option_index].name);
            if (optarg)
                printf(" with arg %s", optarg);
            printf("\n");
            break;

        case 'd':
            conf.daemonize = 1;
            conf.verbosity = VB_QUIET;
            break;
            
        case 's':
            if (conf.server_flag == FLAG_NONE) {
                conf.server_flag = FLAG_SERVER;
            } else {
                fprintf(stderr, "You can only set one mode, either Server or "
                                "Relay Agent, not both.\n\n");
                usage();
            }
            break;

        case 'r':
            if (conf.server_flag == FLAG_NONE) {
                conf.server_flag = FLAG_RELAY;
            } else {
                fprintf(stderr, "You can only set one mode, either Server or "
                                "Relay Agent, not both.\n\n");
                usage();
            }
            break;

        case 'c':
            if (conf.server_flag == FLAG_SERVER) {
                conf.host = strdup(optarg);
                conf.port = parse_host_str(conf.host);
            } else {
                fprintf(stderr, "You must select Server mode to use -c or "
                                "--listen-client.\n\n");
                usage();
            }
            break;

        case 'l':
            if (conf.server_flag == FLAG_SERVER) {
                conf.host2 = strdup(optarg);
                conf.port2 = parse_host_str(conf.host2);
            } else {
                fprintf(stderr, "You must select Server mode to use -l or "
                                "--listen-relay.\n\n");
                usage();
            }
            break;

        case 'b':
            if (conf.server_flag == FLAG_RELAY) {
                conf.host = strdup(optarg);
                conf.port = parse_host_str(conf.host);
            } else {
                fprintf(stderr, "You must select Relay Agent mode to use -c "
                                "or --server-host.\n\n");
                usage();
            }
            break;
    
        case 't':
            if (conf.server_flag == FLAG_RELAY) {
                conf.host2 = strdup(optarg);
                conf.port2 = parse_host_str(conf.host2);
            } else {
                fprintf(stderr, "You must select Relay Agent mode to use -t "
                                "or --target-host.\n\n");
                usage();
            }
            break;
        case 'h':
            usage();
            break;

        case 'k':
            conf.keepalive = atoi(optarg);
            break;
            
        case 'v':
            if (conf.daemonize == 0) {
                conf.verbosity = VB_VERBOSE;
            }
            break;
            
        case 'q':
            conf.verbosity = VB_QUIET;
            break;
            
        case '?':
            /* FALL THROUGH */
        default:
            usage();
        }
    }

    /* Print any remaining command line arguments (not options). */
    if (optind < argc) {
        usage();
    }

    if (conf.daemonize == TRUE) {
        pid = fork();

        if (pid > 0) { /* parent */
            exit(0);
        } else if (pid < 0) { /* error */
            perror("fork");
            exit(1);
        } /* child */
        setsid();

        close(0); close(1); close(2);
    }
    
    signal(SIGTERM, clean_exit);
    signal(SIGINT, clean_exit);
    signal(SIGQUIT, clean_exit);
    signal(SIGHUP, clean_exit);
    
    if (conf.server_flag == FLAG_SERVER) {
        if (conf.port >= 0 && conf.port2 >= 0) {
            server(conf.host, conf.port, conf.host2, conf.port2);
        } else {
            fprintf(stderr, "Two ports, port1 (-l and --client-port) and "
                            "port2 (-L and --ra-port) must be "
                            "provided.\n\n");
            usage();
        }
    } else if (conf.server_flag == FLAG_RELAY) {
        if (conf.host != NULL && conf.host2 != NULL) {
            relay_agent(conf.host, conf.port, conf.host2, conf.port2);
        } else {
            fprintf(stderr, "Both a server (-b and --server-host) and a "
                            "target (-t and --target-host)\nmust be "
                            "specified.\n\n");
            usage();
        }
    } else {
        usage();
    }

    exit(0);
}

void
usage(void)
{

    fprintf(stderr, "Usage: revinetd -s -c HOST1:PORT1 -l HOST2:PORT2 [-d]"
                    "\n"
                    "       revinetd -r -b HOST1:PORT1 -t "
                    "HOST2:PORT2 [-d] [-k seconds]\n"
                    "\n"
                    "Global arguments:\n"
                    "  -s  --server\t\tRuns in server mode\n"
                    "  -r  --relay-agent\tRuns as the relay agent\n"
                    "  -d  --daemonize\tRuns as a daemon (implies -q)\n"
                    "  -h  --help\t\tPrints this message and exit\n"
                    "  -q  --quiet\t\tSurpress all messages\n"
                    "  -v  --verbose\t\tIncrease verbiosity\n"
                    "\n"
                    "Server arguments:\n"
                    "  -c HOST1:PORT1 --client-port=HOST1:PORT1\tListen for a client on IP and Port\n"
                    "  -l HOST2:PORT2 --ra-port=HOST2:PORT2    \tListen for a relay agent on IP and Port\n"
                    "Relay agent arguments:\n"
                    "  required arguments:\n"
                    "  -b HOST1:PORT1 --server-host=HOST1:PORT1\tRevinetd server "
                    "host and port\n"
                    "  -t HOST2:PORT2 --target-host=HOST2:PORT2\tThe target host "
                    "and port\n\n"
                    "  optional arguments:\n"
                    "  -k seconds --keep-alive=seconds\tKeep-alive signal interval\n"
                    "                                 \tdefaults to 180 seconds\n\n");

    clean_exit(SIGQUIT);
}

void
clean_exit(int sig)
{
    OpenSockets *open_sock;
    
    /* Close all registered sockets. */
    open_sock = conf.open_sock;
    while (open_sock != NULL) {
        if (shutdown(open_sock->sock, SHUT_RDWR) == -1) {
            perror("shutdown");
        }
        close(open_sock->sock);
        open_sock = open_sock->next;
    }
    
    exit(0);
}

unsigned short
parse_host_str(char *host_str)
{
    unsigned short port;
    int j = 0;
    
    /* Skip any leading spaces. */
    while (isspace(*host_str)) { host_str++; }
    
    /* Loop through the line until we reach a ':'. Exit with errors if we
       stumble on an illegal character. */
    while (*(host_str + j) != ':') {
        if (*(host_str + j) == '\0' || isspace(*(host_str + j))) {
            fprintf(stderr, "Invalid hostname format.\n");
            usage();
        }
        j++;
    }

    /* Insert a NUL byte at the position of the ':' to delimitate the
       hostname. And then skip ahead one byte. */
    *(host_str + j++) = '\0';
   
    /* Grab the port. */
    port = (unsigned short)atoi(host_str + j);

    return port;
}

void
init_conf(void)
{

    memset(&conf, 0, sizeof(Conf));
    conf.port = -1;
    conf.port2 = -1;
    conf.verbosity = VB_NORMAL;
    conf.keepalive = 180L;
}
